src/ducc0/infra/threading.cc

```diff
@@ -607,7 +619,39 @@ void Distribution::thread_map(std::function<void(Scheduler &)> f)
   // threads, which executes everything sequentially on its own thread,
   // automatically prohibiting nested parallelism.
   auto pool = get_active_pool();
-  for (size_t i=0; i<nthreads_; ++i)
+
+#if 1  // Hierarchical submission
+
+  latch counter(nthreads_);
+  // distribute work to helper threads, in a recursive fashion
+  std::function<void(size_t, size_t)> new_f = [this, &f, &new_f, &counter, &ex, &ex_mut, pool](size_t istart, size_t step) {
+    try
+      {
+      ScopedValueChanger changer(in_parallel_region, true);
+      ScopedUseThreadPool guard(*pool);
+      for(; step>0; step>>=1)
+        if(istart+step<nthreads_)
+          pool->submit([this, &f, &new_f, &counter, &ex, &ex_mut, pool, istart, step]()
+            {new_f(istart+step, step>>1);});
+      MyScheduler sched(*this, istart);
+      f(sched);
+      }
+    catch (...)
+      {
+      LockGuard lock(ex_mut);
+      ex = std::current_exception();
+      }
+    counter.count_down();
+    };
+
+  size_t biggest_step=1;
+  while (biggest_step*2<nthreads_) biggest_step<<=1;
+  new_f(0, biggest_step);
```

Comment:
The problem I see with executing on a thread outside of the thread pool is that you potentially create over-subscription issues where more threads are active than there are cores on the machine. I guess maybe you could disable a thread pool worker whenever a calling thread replaces them, though the additional synchronization to achieve that might negate any benefit.